File: frontend/src/use-cases/MarkAllNotificationsAsRead.ts
// src/use-cases/MarkAllNotificationsAsRead.ts
/**
 * Mark All Notifications As Read Use Case
 * 
 * @description Handles the business logic for marking all user notifications as read
 * @layer Application
 * @pattern Use Case Pattern
 * @dependencies Domain entities, Repository interfaces
 * 
 * @author Capital Craft Team
 * @created 2025-08-09
 */

import { Result } from '../entities/Notification';

/**
 * Repository interface for bulk notification operations
 * Following SOLID principles - Interface Segregation
 */
export interface IBulkNotificationRepository {
  markAllAsRead(userId: string): Promise<Result<{ markedCount: number }>>;
}

/**
 * Mark All Notifications As Read Use Case
 * 
 * @description Encapsulates business logic for bulk marking notifications as read
 * @example
 * ```typescript
 * const useCase = new MarkAllNotificationsAsReadUseCase(repository);
 * const result = await useCase.execute('user-123');
 * if (result.success) {
 *   console.log(`Marked ${result.data.markedCount} notifications as read`);
 * }
 * ```
 */
export class MarkAllNotificationsAsReadUseCase {
  constructor(
    private readonly repository: IBulkNotificationRepository
  ) {}

  /**
   * Execute the mark all notifications as read use case
   * 
   * @param userId - User identifier
   * @returns Promise<Result<{ markedCount: number }>>
   * 
   * @throws {Error} When userId is invalid
   * @throws {Error} When repository operation fails
   */
  async execute(userId: string): Promise<Result<{ markedCount: number }>> {
    // Domain validation
    if (!userId || typeof userId !== 'string' || userId.trim().length === 0) {
      return {
        success: false,
        error: 'Invalid userId provided',
        code: 'INVALID_USER_ID'
      };
    }

    try {
      // Delegate to repository layer
      const result = await this.repository.markAllAsRead(userId.trim());
      
      // Business logic: Log successful bulk operation
      if (result.success && result.data) {
        console.log(`✅ Marked ${result.data.markedCount} notifications as read for user: ${userId}`);
      }
      
      return result;
      
    } catch (error) {
      console.error('❌ Mark all notifications as read failed:', error);
      
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error occurred',
        code: 'BULK_OPERATION_FAILED'
      };
    }
  }
}

/**
 * Factory function for creating use case instances
 * Following Dependency Injection pattern
 */
export const createMarkAllNotificationsAsReadUseCase = (
  repository: IBulkNotificationRepository
): MarkAllNotificationsAsReadUseCase => {
  return new MarkAllNotificationsAsReadUseCase(repository);
};

File: backend/app/use_cases/mark_all_notifications_as_read.py
"""
Mark All Notifications As Read Use Case

@description Handles the business logic for marking all user notifications as read
@layer Application  
@pattern Use Case Pattern
@dependencies Domain entities, Repository interfaces

@author Capital Craft Team
@created 2025-01-15
"""
from typing import List
from ..core.entities.notification import Notification
from ..core.interfaces.notification_repository import NotificationRepository


class MarkAllNotificationsAsReadUseCase:
    """
    Business logic for marking all notifications as read for a user
    
    @description Bulk operation to mark all unread notifications as read
    @layer Application
    @pattern Use Case Pattern
    
    Features:
    - Bulk update operation
    - Returns count of updated notifications
    - Skips dismissed notifications
    - Optimized for performance
    """
    
    def __init__(self, notification_repository: NotificationRepository):
        """
        Initialize use case with repository dependency
        
        @param notification_repository Repository for notification persistence
        """
        self.notification_repository = notification_repository
    
    async def execute(self, user_id: str) -> int:
        """
        Execute the mark all as read use case
        
        @param user_id User identifier
        @returns Number of notifications marked as read
        
        @example
        ```python
        use_case = MarkAllNotificationsAsReadUseCase(repository)
        count = await use_case.execute("demo")
        print(f"Marked {count} notifications as read")
        ```
        """
        
        # Use repository's optimized bulk operation
        marked_count = await self.notification_repository.mark_all_as_read(user_id)
        
        return marked_count
